wallet-new/src/Cardano/Wallet/Client.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StrictData #-}
+
+-- | This module defines a client for the Cardano new-wallet.
+module Cardano.Wallet.Client
+    ( -- * The abstract client
+      WalletClient(..)
+    , getWalletIndex
+    , Resp
+    , hoistClient
+    , liftClient
+    -- * The type of errors that the client might return
+    , ClientError(..)
+    , WalletError(..)
+    , ServantError(..)
+    -- * Reexports
+    , module Cardano.Wallet.API.V1.Types
+    , module Cardano.Wallet.API.V1.Parameters
+    , module Cardano.Wallet.API.Request.Pagination
+    ) where
+
+import           Universum
+
+import           Servant.Client (ServantError (..))
+
+import           Cardano.Wallet.API.Request.Pagination
+import           Cardano.Wallet.API.Response
+import           Cardano.Wallet.API.V1.Errors
+import           Cardano.Wallet.API.V1.Parameters
+import           Cardano.Wallet.API.V1.Types
+import qualified Pos.Core as Core
+
+-- | A representation of a wallet client parameterized over some effect
+-- type @m@.
+--
+-- The record fields serve as the API to the wallet client. Note that the
+-- field 'getAddressIndex' has the type:
+--
+-- @
+-- 'getAddressIndex'
+--     :: 'WalletClient' m
+--     -> m ('Either' 'WalletError' ('WalletResponse' ['Address']))
+-- @
+--
+-- Other functions may be defined in terms of this 'WalletClient' -- see
+-- 'getWalletIndex' as a convenience helper for 'getWalletIndexPaged'.
+data WalletClient m
```

Comment:
Ah, interesting! Eventually you settled for "Option 1" 😛

Reply:
I have used this design before and it is *very* easy to swap out implementations without requiring a ton of `mtl` boilerplate -- passing values is easier than wrangling type classes sometimes :D